File: 0152. Maximum Product Subarray.java
// Medium
// Array
// O(n)
// https://leetcode.cn/problems/maximum-product-subarray/

class Solution {
    public int maxProduct(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }

        int max = nums[0], curMax = nums[0], curMin = nums[0];

        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];

            // 如果當前是負數，也有可能負數*負數成為一個更大的數
            // 那curMin和curMax要互換一下
            if (num < 0) {
                int temp = curMax;
                curMax = curMin;
                curMin = temp;
            }

            // 計算當前元素的最大值和最小值
            curMax = Math.max(num, curMax * num);
            curMin = Math.min(num, curMin * num);

            // 更新全局最大值
            max = Math.max(max, curMax);
        }
        return max;
    }
}

File: 0001-0500/0152. Maximum Product Subarray.py
# Time:O(n), Space:O(1)

from typing import List

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        cur_max = cur_min = res = nums[0]

        for i in range(1, len(nums)):
            num = nums[i]

            if num < 0:
                cur_max, cur_min = cur_min, cur_max

            cur_max = max(num, cur_max * num)
            cur_min = min(num, cur_min * num)

            res = max(res, cur_max)

        return res
